reference/5.1/Microsoft.PowerShell.Core/About/about_Member-Access_Enumeration.md

```diff
@@ -0,0 +1,283 @@
+(Get-Service -Name event*).ForEach({ $_.DisplayName })
+(Get-Service -Name event*).DisplayName
+```
+
+```Output
+Windows Event Log
+COM+ Event System
+
+Windows Event Log
+COM+ Event System
+
+Windows Event Log
+COM+ Event System
+```
+
+When you use the member-access operator on any object, regardless of whether it
+is a scalar or a collection, if the specified member exists on that object, the
+member is invoked. For property members, the operator returns the value of that
+property. For method members, the operator calls that method on the object.
+
+When you use the member-access operator on a collection object that doesn't
+have the specified member, PowerShell automatically enumerates the items in
+that collection and uses the member-access operator on each enumerated item.
+
+During member-access enumeration for a property, the operator returns the value
+of the property for each item that has that property. If no items have the
+specified property, the operator returns `$null`.
+
+During member-access enumeration for a method, the operator attempts to call the
+method on each item in the collection. If any item in the collection does does
+not have the specified method, the operator returns the **MethodNotFound**
+exception.
+
+> [!WARNING]
+> During member-access enumeration for a method, the method is called on each
+> item in the collection. If the method you are calling a method make changes,
+> the changes are made for every item in the collection. If an error occurs
+> during enumeration, the method is called only the items enumerated before the
+> error. For additional safety, consider manually enumerating the items and and
+> explicitly handling any errors.
+
+The following examples detail the behavior of the member-access operator under
+all possible scenarios.
+
+## Accessing members of a scalar object
+
+When you use the member access operator on a scalar object that has the member,
+the command returns the value of the property or output of the method for that
+object.
+
+```powershell
+$Scalar = 'abc'
+$Scalar.Length
+$Scalar.ToUpper()
+```
+
+```Output
+3
+
+ABC
+```
+
+When you use the member access operator on a scalar object that does not have
+the member, the command returns `$null` if you specify a property or a
+**MethodNotFound** error if you specify a method.
+
+```powershell
+$Scalar = 'abc'
+$Scalar.DoesNotExist
+$null -eq $Scalar.DoesNotExist
+$Scalar.DoesNotExist()
+```
+
+```Output
+
+
+True
+
+Method invocation failed because [System.String] does not contain a method
+named 'DoesNotExist'.
+At line:1 char:1
++ $Scalar.DoesNotExist()
++ ~~~~~~~~~~~~~~~~~~~~~~
+    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
+    + FullyQualifiedErrorId : MethodNotFound
+```
+
+## Accessing members of a collection object
+
+When you use the member-access operator on a collection object that has the
+member, it always returns the property value or method result for the
+collection object.
```

Comment:
I may be misremembering, but I _think_ this behaviour has a notable exception; hashtables always prefer a key with a matching name to a property on the hashtable itself:

```ps1
$ht = @{ Keys = 1; OtherStuff = $true }
 # Outputs `1` and not `@(Keys, OtherStuff)`
$ht.Keys

Reply:
Great catch! I'll update the docs with another example to clarify this exception.

Reply:
Good catch. We will add it.

Reply:
[Hashtables are IEnumerable](https://docs.microsoft.com/en-us/dotnet/api/system.collections.hashtable?view=net-6.0).

PowerShell is the only common tool that doesn't consider them enumerable. In C#, a pattern like `foreach (KeyValuePair kvp in someHashtable) { ... }` is valid code, unlike in PowerShell.

Many folks with prior .NET experience may expect dictionary types to be considered enumerable / collections the same way .NET tends to treat them, so it's a good idea to call out things where there are differences to how PS normally treats other collections.

Reply:
Interesting that it's not ICollection 🤔  Thanks for taking the time to chase it down. 🙂